pkg/cloudprovider/aws.go

```diff
@@ -1,37 +1,338 @@
 
-func (a *AWS) WaitForResponse(interface{}) error {
-	return nil
+func (a *AWS) WaitForResponse(requestObj interface{}) error {
+	awsWaitInput, ok := requestObj.(AWSWaitInput)
+	if !ok {
+		return fmt.Errorf("error decoding AWS requestObj, object not of type: AWSWaitInput %#v", requestObj)
+	}
+	var ec2IPFilter string
+	sampleIP := *awsWaitInput.ips[0]
+	if utilnet.IsIPv6String(sampleIP) {
+		ec2IPFilter = awsIPv6FilterKey
+	} else {
+		ec2IPFilter = awsIPv4FilterKey
+	}
+	err := a.client.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{awsWaitInput.instanceID},
+		Filters: []*ec2.Filter{
+			&ec2.Filter{
+				Name:   &ec2IPFilter,
+				Values: awsWaitInput.ips,
+			},
+		},
+	})
+	return err
 }
 
 func (a *AWS) GetNodeEgressIPConfiguration(node *corev1.Node) ([]*NodeEgressIPConfiguration, error) {
-	return nil, nil
+	configs := []*NodeEgressIPConfiguration{}
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving instance associated with node, err: %v", err)
+	}
+	instanceV4Capacity, instanceV6Capacity, err := a.getInstanceCapacity(instance)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving the instance capacities, err: %v", err)
+	}
+	for _, networkInterface := range a.getNetworkInterfaces(instance) {
+		config := &NodeEgressIPConfiguration{
+			Interface: *networkInterface.NetworkInterfaceId,
+		}
+		v4Subnet, v6Subnet, err := a.getSubnet(networkInterface)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving the network interface subnets, err: %v", err)
+		}
+		config.IFAddr = ifAddr{}
+		if v4Subnet != nil {
+			config.IFAddr.IPv4 = v4Subnet.String()
+		}
+		if v6Subnet != nil {
+			config.IFAddr.IPv6 = v6Subnet.String()
+		}
+		capV4, capV6 := a.getCapacity(instanceV4Capacity, instanceV6Capacity, networkInterface)
+		config.Capacity = capacity{
+			IPv4: capV4,
+			IPv6: capV6,
+		}
+		configs = append(configs, config)
+		break
+	}
+	return configs, nil
+}
+
+func (a *AWS) getSubnet(networkInterface *ec2.InstanceNetworkInterface) (*net.IPNet, *net.IPNet, error) {
+	describeOutput, err := a.client.DescribeSubnets(&ec2.DescribeSubnetsInput{
+		SubnetIds: []*string{networkInterface.SubnetId},
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("error: cannot list ec2 subnets, err: %v", err)
+	}
+	if len(describeOutput.Subnets) > 1 {
+		return nil, nil, fmt.Errorf("error: multiple subnets found for the subnet ID: %s", *networkInterface.SubnetId)
+	}
+	var v4Subnet, v6Subnet *net.IPNet
+	subnet := describeOutput.Subnets[0]
+	if *subnet.CidrBlock != "" {
+		_, subnet, err := net.ParseCIDR(*subnet.CidrBlock)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error: unable to parse IPv4 subnet, err: %v", err)
+		}
+		v4Subnet = subnet
+	}
+
+	// I don't know what it means to have several IPv6 CIDR blocks defined for
+	// one subnet, specially given that you can only have one IPv4 CIDR block
+	// defined...¯\_(ツ)_/¯
+	// Let's just pick the first.
```

Comment:
I'm not sure that's safe.

I guess really you want the IPv6 subnet that contains the IPv6 node IP, right?

Reply:
I've been thinking about this and I am not sure that makes more sense. What if the node has multiple IPv6 addresses? If it has one it probably has one or multiple CIDR blocks. 

Now, if there are multiple CIDR blocks we probably need to redefine the annotation, since it currently only assumes one-to-one relation between IP family subnets and NIC. Since all of this is very speculative, conceptual and strange at this point (since I am not able to bring up a IPv6 / dual-stack cluster on AWS and verify this): I think it's fine to just take the first IPv6 CIDR block if there are multiple defined, and not redefine the annotation for now.